aten/src/ATen/native/sparse/SoftMax.cpp

```diff
@@ -0,0 +1,467 @@
+#include <ATen/ATen.h>
+#include <ATen/Config.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/SparseTensorUtils.h>
+#include <ATen/Parallel.h>
+#include <ATen/NamedTensorUtils.h>
+#include <map>
+
+namespace at {
+namespace native {
+namespace {
+
+template <typename iscalar_t>
+iscalar_t get_nvalues(const std::vector<iscalar_t>& sizes, const int64_t sparse_dim) {
+  /* Return the number of elements of sparse values entries.
+   */
+  iscalar_t nvalues = 1;
+  for (auto it = sizes.begin() + sparse_dim; it != sizes.end(); ++it) {
+    nvalues *= *it;
+  }
+  return nvalues;
+}
```

Comment:
nit: Also the description is not clear to me. Looks like you compute the number of elements from the specified dim to the end?

Reply:
Yes, in sparse tensors the dimensions are "sorted": the first part of `sizes` corresponds to sparse dimensions and the rest to dense dimensions. The `get_nvalues` computes the number of entries of the dense part of the sparse tensor.

Reply:
Updated the description.